lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,240 @@
+    #
+    # ## Composing state machines
+    #
+    # Multiple state machines can be composed into a single unified state machine with the {#compose} method.
+    # When state machines are composed the state names and transitions are namespaced based on the originating
+    # state machine.
+    #
+    # ### Final states
+    #
+    # States that were marked as final are converted to intermediate states so that transitions
+    # can be added between states.
+    #
+    # @example
+    #   Puppet::StateMachine.build("Example Puppet runner") do |m|
+    #     m.start_state(:start)
+    #
+    #     m.state(:start,
+    #       action: -> { Puppet.notice("Starting Puppet agent") },
+    #       event: ->(result) { :create_lockfile },
+    #       transitions: {create_lockfile: :create_lockfile})
+    #
+    #     m.state(:create_lockfile,
+    #       action: -> { Puppet::Util::Lockfile.new("/var/lock/puppet.lock").lock },
+    #       event: ->(has_lock) { has_lock ? :got_lock : :already_locked },
+    #       transitions: {
+    #         got_lock: :run_agent,
+    #         already_locked: :cancel_run,
+    #       })
+    #
+    #     m.state(:cancel_run,
+    #       type: :error,
+    #       action: -> { Puppet.error("Cancelling run.") }
+    #
+    #     m.state(:run_puppet,
+    #       action: -> { Puppet::Agent.run },
+    #       event: ->(exit_code) { exit_code == 0 ? :success : :failure },
+    #       transitions: {
+    #         success: :shutdown_success,
+    #         failure: :shutdown_failure
+    #       })
+    #
+    #     m.state(:shutdown_success,
+    #       type: :final,
+    #       action: -> do
+    #         Puppet::Util::Lockfile.new("/var/lock/puppet.lock").unlock
+    #         Puppet.notice("Puppet run complete")
+    #       end
+    #
+    #     m.state(:shutdown_failure,
+    #       type: :error,
+    #       action: -> do
+    #         Puppet::Util::Lockfile.new("/var/lock/puppet.lock").unlock
+    #         Puppet.error("Puppet run failed")
+    #       end
+    #   end
+    #
+    #
+    class MachineBuilder
+
+      # @param machine_name [String] The name of the machine to build.
+      def initialize(machine_name)
+        @machine_name = machine_name
+        @states = {}
+        @start_state = nil
+      end
+
+      # Build and verify a new state machine.
+      # @return [Puppet::StateMachine]
+      def build(&blk)
+        yield self
+        verify!
+        Puppet::StateMachine.new(@machine_name, @states, @start_state)
+      end
+
+      # Compose multiple state machines into a a single state machine.
+      #
+      # @example
+      #   machine.compose({machine1: m1, machine2: m2}) do |m|
+      #     m.start_state([:m1, :start])
+      #
+      #     m.transition(
+      #       source: [:m1, :final],
+      #       target: [:m2, :start],
+      #       on: :'*final*')
+      #
+      #     m.final([:m2, :final])
+      #   end
+      def compose(machines, &blk)
```

Comment:
The states are copied from the original state machine, namespaced, and then placed in the new state machine. For the case where m1 has states on and off, and m2 has multiple states for that on state, you would model this with the following.

```ruby
on_machine = Puppet::StateMachine.build("On states") do |m|
  m.start_state(:initializing)

  m.state(:initializing,
    action: -> { Puppet.notice("initializing") },
    event: ->(result) { :initialized },
    transitions: {initialized: :on}) # Assuming that initialization can't fail

  m.state(:operative,
    action: -> { Puppet.notice("operating") },
    event: ->(result) { result == :halt ? :shutdown : :run },
    transitions: {
      halt: :shutting_down,
      run: :operative # Self referential transition
    })

  m.state(:shutting_down,
    action: -> { Puppet.notice("Shutting down") },
    event: ->(result) { :halted },
    transitions: {halted: :final})

  m.state(:final,
    type: :final,
    action: -> { Puppet.notice("Halted") })
end

composed = Puppet::StateMachine.build(
  "Composed machine",
  {on: on_machine)
) do |m|
  m.start_state(:start)

  m.state(:start,
          action: -> { Puppet.notice("starting composed machine") },
          event: ->(_) { :turn_on },
          transitions: {turn_on: [:on, :initializing]})

  m.state(:final
          type: :final
          action: -> { Puppet.notice("finishing composed machine") })

  m.transition(
    source: [:on, :final],
    target: :final,
    on: Puppet::StateMachine::FINAL)
end

```